### ISSUE TITLE
Support patching local classes

### ISSUE THREAD

**Issue by donkirkby** [donkirkby/live-py-plugin#83]
When I try to patch a local class in live coding mode, it fails. Here's an example:

```
# patch_example.py
from mock.mock import patch, call


class Foo(object):
    def foo(self, n):
        return n + 3

f = Foo()
n1 = f.foo(10)
with patch('__main__.Foo.foo', side_effect=Foo.foo, autospec=True) as foo_mock:
    n2 = f.foo(20)
    n3 = f.foo(30)
n4 = f.foo(40)

assert n1 == 13, n1
assert n2 == 23, n2
assert n3 == 33, n3
assert n4 == 43, n4
calls = foo_mock.mock_calls
expected_calls = [call(f, 20), call(f, 30)]
assert calls == expected_calls, calls
print('Done.')
```

The error is, "ImportError: No module named Foo". The problem is that live code doesn't run as module `__main__`, it uses the file name as the module name. You can work around it in a couple of ways. One way is to change the patch command:

```
with patch('patch_example.Foo.foo', side_effect=Foo.foo, autospec=True) as foo_mock:
```

The problem with that is that now it won't work when you run it as `__main__`. The other work around is to copy the class you want to patch into the main module.

```
sys.modules['__main__'].Foo = Foo
```

To fix this, we might be able to delete all the global objects from `__main__`, and then run the instrumented code against the existing module. Be careful to check if you can still call unit tests to drive the instrumented code.

Related code from [coverage] package.

[coverage]: https://github.com/nedbat/coveragepy/blob/coverage-4.2/coverage/execfile.py#L133

**Comment by donkirkby** [comment]
If we end up adding the live testing button for issue #78, then this issue could be solved by using the `__name__` variable:
* When you run a script normally, `__name__ == '__main__'`.
* When you run it in live coding mode, `__name__ == '__live_coding__'`.
* When you import from another module, like a unit test, `__name__` is the module's file name without the extension. This would even be true in live testing mode.

If that all works, then the patch command would look like this:

    with patch(__name__ + '.Foo.foo', side_effect=Foo.foo, autospec=True) as foo_mock:

**Comment by donkirkby** [comment]
Related problem with pickling:

    # scratch.py        | 
                        | 
    import pickle       | 
    import sys          | 
                        | 
                        | 
    class Foo(object):  | 
        pass            | 
                        | 
    f = Foo()           | 
    f.foo = 3           | f.foo = 3 
                        | 
    s = pickle.dumps(f) | PicklingError: Can't pickle <class '__live_coding__.Foo'>: it's not found as __live_coding__.Foo 

The problem is that it tries to look up the `Foo` class in the `__live_coding__` module, but the code is loaded as the `scratch` module. Here's a workaround for now:

    # scratch.py                                            | 
                                                            | 
    import pickle                                           | 
    import sys                                              | 
                                                            | 
                                                            | 
    class Foo(object):                                      | 
        pass                                                | 
                                                            | 
    f = Foo()                                               | 
    f.foo = 3                                               | f.foo = 3 
                                                            | 
    sys.modules['__live_coding__'] = sys.modules['scratch'] | 
                                                            | 
    s = pickle.dumps(f)                                     | s = "ccopy_reg\n_reconstructor\np0\n(c__live_coding__\nFoo\np1\nc__builtin__\nobject\np2\nNtp3\nRp4\n(dp5\nS'foo'\np6\nI3\nsb."

**Comment by donkirkby** [comment]
This was fixed by the changes for #79.
